Adds wide gamut framework breakage notice. (#11082)
flutter/website: src/release/breaking-changes/wide-gamut-framework.md

---
title: Migration guide for wide gamut Color
description: Description on Color changes for wide gamut support and migration
             instructions.
---

## Summary

The API for the [`Color`][] class in `dart:ui` is changing to support wide gamut
colors.

> ## Context

Reviewer: I would expect to see 1-2 lines of what wide gamut is, and why it's impactful/important to Flutter users. Remember that the audience for these docs are often TLs or leads on other teams, and they might be asked by _their_ managers/TLs/PMs about why they are spending time on X.

(1-2 lines pointing to Wikipedia or elsewhere is more than sufficient)

Author: Done.  I put it in `Context`.

> Wide gamut color support was [added to the renderer][] previously. That support

Reviewer: Phrasing.

Instead, consider:

> The Flutter engine [already supports Wide gamut color][], and the support is now being added [to the framework][].

Author: Done

is now being plumbed up [to the Flutter Framework][] so those colors can be used
by UIs.

## Description of change

Changes to [`Color`][]:

 1. Adds a enum field which will specify its [`ColorSpace`][].

>  1. Adds API to use normalized floating-point color components.

Reviewer: > to use normalized floating-point color components

Perhaps:

> to use normalized floating-point color components (0.0 to 1.0)

Author: I don't want to say 0 to 1 because it's technically not true.  The extended srgb colorspace is normalized across sRGB, not its limits.

 1. Removes API that uses 8-bit unsigned integer color components that can lead
    to data loss.

Changes to [`ColorSpace`][]:

 1. `displayP3` will be added.

## Migration guide

### 8-bit unsigned integer constructors

Constructors like `Color.fromARGB` will remain unchanged and have continued
support. If one wants to take advantage of Display P3 colors they will have to
use the new `Color.from` constructor that takes normalized floating-point

> color components.

Reviewer: It would be nice to add a dart snippet showing examples here:

```dart
Color.fromARGB(0xFF, 0xFF, 0xFF, 0xFF)
Color.from(1.0, 1.0, 1.0, 1.0)
```

> ### Implementors of `Color`

Reviewer: Do we want to add something about (perhaps) `Color` no longer being implementable in the far future?

Author: Added a note.

---
title: Migration guide for wide gamut Color

> description: Description on Color changes for wide gamut support and migration
>              instructions.

Reviewer 2: ```suggestion
description: >-
  Changes to support wide gamut color and migration instructions.
```

---

## Summary

The API for the [`Color`][] class in `dart:ui` is changing to support wide gamut

> colors.

Author: Done, that link is a bit off since it talks about a specific color space.  I linked https://en.wikipedia.org/wiki/RGB_color_spaces

The iOS devices Flutter targets support rendering a larger array of colors,
specifically the [DisplayP3][] color space. After this change, the Flutter
framework will be able to render all of those colors on iOS Impeller and `Color`
will be better prepared for the future addition of color spaces or changes to
color component bit depth.

## Description of change

Changes to [`Color`][]:

 1. Adds a enum field which will specify its [`ColorSpace`][].
 1. Adds API to use normalized floating-point color components.
 1. Removes API that uses 8-bit unsigned integer color components that can lead
    to data loss.

Changes to [`ColorSpace`][]:

 1. `displayP3` will be added.

## Migration guide

### 8-bit unsigned integer constructors

Constructors like `Color.fromARGB` will remain unchanged and have continued
support. If one wants to take advantage of Display P3 colors they will have to
use the new `Color.from` constructor that takes normalized floating-point
color components.

```dart
// Before
final magenta = Color.fromARGB(0xff, 0xff, 0x0, 0xff);
// After
final magenta = Color.from(alpha: 1.0, red: 1.0, green: 0.0, blue: 1.0)
```

> ### Implementors of `Color`

Author: `implements` is a specific Dart concept and keyword.  How about "Class that `implements` `Color`"?

> There are new methods being added to `Color` so any class that `implements`
> Color will break and have to implement the new methods like `Color.a`, and
> `Color.b`. Ultimately implementors should be migrated to take advantage of the
> new API. In the short-term, these methods can easily be implemented without
> changing the underlying structure of your class.

Reviewer 2: ```suggestion
Because this change adds new methods to `Color`, any app that implements `Color`
will break and must be updated to implement the new methods, like `Color.a` and `Color.b`.
In the short-term, you can do this without changing the underlying structure of your class.
However, in the long term, your reimplementation of `Color` should take advantage of the new API.
```

Author: I think `implements` should be treated as a specific Dart concept and not talked about generally.  It means that you are making a class that guarantees will match another classes public interface.

For example:

```dart
class Foo implements Color {
  int _red;

  @override
  double get r => _red * 255.0;
}
```

> {{site.alert.note}}
>   The hope is to eventually make Color `sealed`. Now might be a good opportunity
>   to stop implementing `Color`. Instead of inheritance, switch to composition.
> {{site.alert.end}}

Reviewer 2: ```suggestion
:::note
We plan to eventually lock the `Color` class down and make it `sealed`. Now might be a good opportunity to switch from [inheritance to composition][] and stop reimplementing `Color`.
:::

[inheritance to composition]: https://en.wikipedia.org/wiki/Composition_over_inheritance
```

Author: I don't see another instance of the `:::note` syntax anywhere in the repo.  Are you sure it's available here?

Reviewer 2: Yes, I am 100% sure.

Author: done

### Color space support

> Clients that use `Color` and perform any sort of calculation on the color
> components should now be checking the color space component before doing any
> such calculation. `Color` has a new method `Color.withValues` that can be used
> to perform color space conversions to help with that.

Reviewer 2: ```suggestion
Clients that use `Color` and perform any sort of calculation on the color
components should now first check the color space component before performing calculations. 
To help with that, you can use the new `Color.withValues` method
to perform color space conversions.
```

For example:

```dart
// Before
double redRatio(Color x, Color y) => x.red / y.red;

// After
double redRatio(Color x, Color y) {
  final xPrime = x.withValues(colorSpace: ColorSpace.extendedSRGB);
  final yPrime = y.withValues(colorSpace: ColorSpace.extendedSRGB);
  return xPrime.r / yPrime.r;
}
```

> ### Color component accessors

Reviewer 2: ```suggestion
### Access color components
```

> For users of `Color`, if the components are ever accessed the code should be
> migrated to take advantage of the floating-point components. In the short term
> one can easily scale the components themselves.

Reviewer 2: "In the short term"? This feels unfinished. What happens in the long term? 

```suggestion
If your app ever accesses a `Color` component, consider taking
advantage of the floating-point components. In the short term,
you can easily scale the components themselves.
```

Author: The long term is the first phrase.  That's the preferred course of action.  The short term suggestion is there if for some reason that's not an easy change.

Reviewer 2: I still prefer the suggestion.  :)

Author: sgtm, done

---
title: Migration guide for wide gamut Color
description: >-
  Changes to support wide gamut color and migration instructions.
---

## Summary

The API for the [`Color`][] class in `dart:ui` is changing to support [wide
gamut color spaces][].

## Context

The Flutter engine [already supports wide gamut color][] with [Impeller][], and the support is now being added [to the framework][].

The iOS devices that Flutter supports render to a larger array of colors,
specifically in the [DisplayP3][] color space. After this change, the Flutter
framework can render all of those colors on iOS Impeller, and the `Color` class
will be better prepared for future color spaces or changes to
color component bit depth.

## Description of change

Changes to [`Color`][]:

 1. Adds an enum field that specifies its [`ColorSpace`][].
 1. Adds API to use normalized floating-point color components.
 1. Removes API that uses 8-bit unsigned integer color components that can lead
    to data loss.

Changes to [`ColorSpace`][]:

>  1. Add a `displayP3` property.

Reviewer 3: To be consistent with the other list, consider:

```suggestion
 1. Adds a `displayP3` property.
```

color components.

```dart
// Before
final magenta = Color.fromARGB(0xff, 0xff, 0x0, 0xff);
// After
final magenta = Color.from(alpha: 1.0, red: 1.0, green: 0.0, blue: 1.0)
```

### Implementors of `Color`

There are new methods being added to `Color` so any class that `implements`
Color will break and have to implement the new methods like `Color.a`, and
`Color.b`. Ultimately implementors should be migrated to take advantage of the
new API. In the short-term, these methods can easily be implemented without
changing the underlying structure of your class.

For example:

```dart
class Foo implements Color {
  int _red;

  @override
  double get r => _red * 255.0;
}
```

{{site.alert.note}}
  The hope is to eventually make Color `sealed`. Now might be a good opportunity
  to stop implementing `Color`. Instead of inheritance, switch to composition.
{{site.alert.end}}

### Color space support

Clients that use `Color` and perform any sort of calculation on the color
components should now first check the color space component before performing calculations. 
To help with that, you can use the new `Color.withValues` method
to perform color space conversions.

For example:

```dart

> // Before

Reviewer 3: I think this is clear if you read the preceding paragraph, but maybe consider making it overly obvious why this is bad.

Maybe:

```suggestion
// BAD - Wrong if `x` is SRGB and `y` is DisplayP3 
```

Author: Done, I spelled in out in prose below.
